### PR TITLE
[app_dart] Create scheduler service

### PR DESCRIPTION
**Pull request by CaseyHillers**
Refactors logic in `refresh_github_commits` into a service

# Issues

https://github.com/flutter/flutter/issues/76140

# Future work
* Trigger adding tasks from github webhooks
  * Remove the cron method
* Move retry logic into this service

### CODE REVIEW

File: app_dart/lib/src/service/scheduler.dart [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+import '../request_handling/exceptions.dart';
+import '../service/luci.dart';
+import 'datastore.dart';
+
+/// Scheduler service to validate all commits to supported Flutter repositories.
+///
+/// Scheduler responsibilties include:
+///   1. Tracking commits in Cocoon
+///   2. Ensuring commits are validated (via scheduling tasks against commits)
+///   3. Retry mechanisms for taskss
```
Comment (keyonghan) [inline app_dart/lib/src/service/scheduler.dart:28]: nit: typo

File: app_dart/lib/src/service/scheduler.dart [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+
+    await _uploadToBigQuery(commit);
+  }
+
+  /// Return subset of [commits] not stored in Datastore.
+  Future<List<Commit>> _getNewCommits(List<Commit> commits) async {
+    final List<Commit> newCommits = <Commit>[];
+    // Ensure commits are sorted from newest to oldest
+    commits.sort((Commit a, Commit b) => a.timestamp.compareTo(b.timestamp));
+    print(commits);
```
Comment (keyonghan) [inline app_dart/lib/src/service/scheduler.dart:81]: Remove?

File: app_dart/lib/src/service/scheduler.dart [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+    });
+
+    try {
+      await datastore.withTransaction<void>((Transaction transaction) async {
+        transaction.queueMutations(inserts: tasks);
+        await transaction.commit();
+        log.debug('Committed ${tasks.length} new tasks for commit ${commit.sha}');
+      });
+    } catch (error) {
+      log.error('Failed to add commit ${commit.sha}: $error');
```
Comment (keyonghan) [inline app_dart/lib/src/service/scheduler.dart:146]: This is for tasks update.

File: app_dart/lib/src/service/scheduler.dart [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+    }
+
+    await _uploadToBigQuery(commit);
+  }
+
+  /// Return subset of [commits] not stored in Datastore.
+  Future<List<Commit>> _getNewCommits(List<Commit> commits) async {
+    final List<Commit> newCommits = <Commit>[];
+    // Ensure commits are sorted from newest to oldest
+    commits.sort((Commit a, Commit b) => a.timestamp.compareTo(b.timestamp));
```
Comment (keyonghan) [inline app_dart/lib/src/service/scheduler.dart:80]: list is already sorted.

File: app_dart/lib/src/request_handlers/refresh_github_commits.dart [hunk 5]
```diff
@@ -75,186 +72,31 @@ class RefreshGithubCommits extends ApiRequestHandler<Body> {
...
-
-      if (newCommits.isEmpty) {
-        // Nothing to do.
-        continue;
-      }
-      log.debug('Found ${newCommits.length} new commits for branch $branch on GitHub');
-
-      // Save [Commit] to BigQuery and create [Task] in Datastore.
-      await _saveData(newCommits, datastore);
+      final List<Commit> recentCommits = await _getRecentCommits(commits, datastore, branch);
```
Comment (keyonghan) [inline app_dart/lib/src/request_handlers/refresh_github_commits.dart:75]: Move this logic to scheduler as well to save both time and space (processing `RepositoryCommit` directly)

File: app_dart/lib/src/service/scheduler.dart (hunk 4 above)
Comment (CaseyHillers) [reply app_dart/lib/src/service/scheduler.dart:78]: This is mostly for tests. I removed documentation. In the next PR, I want to switch this from a list to a single commit and will be irrelevant.

File: app_dart/lib/src/request_handlers/refresh_github_commits.dart (hunk 5 above)
Comment (CaseyHillers) [reply app_dart/lib/src/request_handlers/refresh_github_commits.dart:75]: I left this code here as in the follow up PR I want to remove it completely (move to a single commit from a webhook). SG?

**Review by keyonghan — APPROVED** [review]
LGTM

**Review by godofredoc — COMMENTED** [review]
Given that this is new functionality should it be implemented as a separate service detached from cocoon?  ideally this should be a microservice with a well defined api, implementing and then migrating later is more complicated because we need to consider backward compatibility, etc.

**Pull request merged**